### PR TITLE
docs: Transaction consensus rules: Coinbase rules

### LINKED ISSUE #3427: Document consensus rules from Zcash spec 7.1.2: Coinbase transactions
- [ ] A transaction with one or more transparent inputs from coinbase transactions MUST have no transparent outputs (i.e. `tx_out_count` MUST be 0). Inputs from coinbase transactions include Founders’ Reward outputs and funding stream outputs.
- [ ] The total value in zatoshi of transparent outputs from a coinbase transaction, minus `vbalanceSapling`, minus `vbalanceOrchard`, MUST NOT be greater than the value in zatoshi of block subsidy plus the transaction fees paid by transactions in this block.
- [ ] A coinbase transaction MUST NOT have any transparent inputs with non-null prevout fields, JoinSplit descriptions, or Spend descriptions.
- [ ] [Pre-Heartwood] A coinbase transaction also MUST NOT have any Output descriptions.
- [ ] [NU5 onward] In a version 5 coinbase transaction, the `enableSpendsOrchard` flag MUST be 0.
- [ ] A coinbase transaction for a block at block height greater than 0 MUST have a script that, as its first item, encodes the block height `height` as follows. For `height` in the range {1..16}, the encoding is a single byte of value `0x50 height`. Otherwise, let `heightBytes` be the signed little-endian representation of `height`, using the minimum nonzero number of bytes such that the most significant byte is `< 0x80`. The length of `heightBytes` MUST be in the range {1..5}. Then the encoding is the length of `heightBytes` encoded as one byte, followed by `heightBytes` itself. This matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the description here is to be considered normative).
- [ ] A transaction MUST NOT spend a transparent output of a coinbase transaction from a block less than 100 blocks prior to the spend. Note that transparent outputs of coinbase transactions include Founders’ Reward outputs and transparent funding stream outputs.
- [ ] A transaction MUST NOT spend an output of the genesis block coinbase transaction. (There is one such zero-valued output, on each of Testnet and Mainnet.)
- [ ] [NU5 onward] The `nExpiryHeight` field of a coinbase transaction MUST be equal to its block height.
- [ ] [Heartwood onward] All Sapling and Orchard outputs in coinbase transactions MUST decrypt to a note plaintext, i.e. the procedure in § 4.19.3 ‘Decryption using a Full Viewing Key (Sapling and Orchard)’ on p. 66 does not return ⊥, using a sequence of 32 zero bytes as the outgoing viewing key. (This implies that before Canopy activation, Sapling outputs of a coinbase transaction MUST have note plaintext lead byte equal to 0x01.)
- [ ] [Canopy onward] Any Sapling or Orchard output of a coinbase transaction decrypted to a note plaintext according to the preceding rule MUST have note plaintext lead byte equal to 0x02. (This applies even during the “grace period” specified in [ZIP-212].)

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We want to make sure all coinbase transaction rules are implemented and quoted in the format we want in the codebase.

If merged this will close https://github.com/ZcashFoundation/zebra/issues/3427

## Solution

Make the needed changes, almost all was just adjusting to the new format as rules docs were already there.

## Review

Anyone can review.

### Reviewer Checklist

- [ ] Consensus rule docs are in a good spot.
- [ ] All the rules in the ticket were added or modified in this PR.
- [ ] No piece of doc comment that i want to keep was deleted.

## Follow Up Work

The other 2 sections of the transaction consensus rules which are in the following issues:

- https://github.com/ZcashFoundation/zebra/issues/3428

### CODE REVIEW

**Comment by oxarbitrage** [comment]
A coinbase transaction for a block at block height greater than 0 MUST have a script that, as its first item, encodes the block height height as follows. For height in the range {1..16}, the encoding is a single byte of value 0x50 height. Otherwise, let heightBytes be the signed little-endian representation of height, using the minimum nonzero number of bytes such that the most significant byte is < 0x80. The length of heightBytes MUST be in the range {1..5}. Then the encoding is the length of heightBytes encoded as one byte, followed by heightBytes itself. This matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the description here is to be considered normative).

I was not able to find the above rule, i am pretty sure we did this(or part of it) but can't remember where.

**Comment by teor2345** [comment]
> I was not able to find the above rule, i am pretty sure we did this(or part of it) but can't remember where.

Sounds like:
https://github.com/ZcashFoundation/zebra/blob/fa071562fd0b891f29143e4cc8d17b34b4b1cf10/zebra-chain/src/transparent/serialize.rs#L68-L70

File: zebra-consensus/src/transaction/check.rs [hunk 1]
```diff
@@ -99,14 +99,20 @@ pub fn has_enough_orchard_flags(tx: &Transaction) -> Result<(), TransactionError
 
 /// Check that a coinbase transaction has no PrevOut inputs, JoinSplits, or spends.
 ///
-/// A coinbase transaction MUST NOT have any transparent inputs, JoinSplit descriptions,
-/// or Spend descriptions.
+/// # Consensus
 ///
-/// In a version 5 coinbase transaction, the enableSpendsOrchard flag MUST be 0.
+/// >  A coinbase transaction MUST NOT have any transparent inputs with non-null prevout fields,
```
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs]: Nit: extra space

```suggestion
/// > A coinbase transaction MUST NOT have any transparent inputs with non-null prevout fields,
```

File: zebra-consensus/src/transaction/check.rs [hunk 2]
```diff
@@ -99,14 +99,20 @@ pub fn has_enough_orchard_flags(tx: &Transaction) -> Result<(), TransactionError
...
+/// > JoinSplit descriptions, or Spend descriptions.
+///
+/// > [NU5 onward] In a version 5 coinbase transaction, the enableSpendsOrchard flag MUST be 0.
 ///
 /// This check only counts `PrevOut` transparent inputs.
 ///
-/// https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+/// > [Pre-Heartwood] A coinbase transaction also MUST NOT have any Output descriptions.
+///
+/// Zebra will not validate this last rule explicitly because we checkpoint until Canopy blocks.
```
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs]: Nit: 'will' makes it sound like Zebra's validation is incomplete.

```suggestion
/// Zebra does not validate this last rule explicitly because we checkpoint until Canopy activation.
```

File: zebra-consensus/src/transaction/check.rs [hunk 3]
```diff
@@ -298,8 +307,12 @@ pub fn coinbase_expiry_height(
...
         None => validate_expiry_height_max(expiry_height, true, block_height, coinbase),
         Some(activation_height) => {
-            // Consensus rule: from NU5 activation, the nExpiryHeight field of a
-            // coinbase transaction MUST be set equal to the block height.
+            // # Consensus
+            //
+            // > [NU5 onward] The nExpiryHeight field of a coinbase transaction MUST be equal
+            // > to its block height.
+            //
+            // https://zips.z.cash/protocol/protocol.pdf#txnconsensus
```
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs:315]: This change will conflict with PR #3401, which changes this code.

Can you move it to a separate PR that is based on #3401?

**Review by teor2345 — COMMENTED** [review]
Looks good, but we should try to avoid a conflict with PR #3401.

File: zebra-consensus/src/transaction/check.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebra-consensus/src/transaction/check.rs:315]: By what i can see #3401 already has 2 conflicts in the same file, so i dont think it will make sense to base anything on that PR right now. On the contrary, it seems #3401 will need to fix conflicts to be up to date. We can just remove the change and leave #3401 handle it or keep it as it is and have #3401 deal with it as well.

Comment (teor2345) [reply zebra-consensus/src/transaction/check.rs:315]: I'm happy to do whatever you think is best here.

**Review by teor2345 — APPROVED** [review]
Looks good, thanks!
